### PR TITLE
Dx 35 cli create publish plugin script

### PR DESCRIPTION
**Pull request by pythonpete32**
`aragon publish` command: publishes the plugin into Aragon OSx protocol and creates a PluginRepo instance containing the plugin’s first version.

[video of command](https://share.cleanshot.com/jsV3NPnG)
[video of failure case](https://share.cleanshot.com/wyTZzD5T)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,7 +16,93 @@
...
+```
+aragon settings set-pk [private-key]
+```
+
+### `aragon settings set-tenderly`
+
+Store a new Tenderly project/apikey configuration.
+
+```
+aragon settings set-tenderly
```
Comment (juliettech13) [inline README.md:94]: "settings set-tenderly"

File: README.md [hunk 2]
```diff
@@ -16,7 +16,93 @@
...
+```
+
+### `aragon settings view`
```
Comment (juliettech13) [inline README.md:97]: "settings view"

File: src/commands/publish/index.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { Command, Option } from 'commander';
+import { publishHandler } from './publishHandler';
+
+export const publish = new Command('publish');
+
+publish
+  .description('publish a Plugin to a PluginRepo')
```
Comment (juliettech13) [inline src/commands/publish/index.ts:7]: important to keep consistent across capitalization here. either all lower cased or all upper cased. let's go for capitalizing all of them.

File: src/commands/publish/publishHandler.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    setupContract,
+    subdomain: `${subdomain}.plugin.dao.eth`,
+    chain: network.name,
+    maintainer,
+  });
+
+  const simulate = options.simulate ?? (await confirmPrompt(strings.SIMULATE_DEPLOYMENT));
+  if (simulate) await simulatePublish(setupContract, network, subdomain, maintainer);
+
+  // confirm publish
```
Comment (juliettech13) [inline src/commands/publish/publishHandler.ts:50]: We don't need this comment here

File: src/lib/schemas.ts [hunk 5]
```diff
@@ -1,26 +1,71 @@
...
+  (value) => {
+    const ethereumAddressPattern = /^0x[a-fA-F0-9]{40}$/;
+    return ethereumAddressPattern.test(value);
+  },
+  {
+    message: 'Invalid Ethereum address',
+  },
+);
+
+// export const AllowedNetworksSchema = z.enum(Object.keys(activeContractsList) as const);
```
Comment (juliettech13) [inline src/lib/schemas.ts:33]: Remove comment

File: src/lib/schemas.ts [hunk 6]
```diff
@@ -1,26 +1,71 @@
...
+  .regex(/^[a-zA-Z0-9-]*$/, 'Tenderly key must only contain numbers and letters.');
 
 export const contractNameSchema = z
   .string()
   .nonempty({ message: 'Contract name cannot be empty.' })
   .regex(/^[a-zA-Z0-9]*$/, {
     message: 'Contract name can only contain numbers and letters.',
   });
+
+export const ethereumAddressSchema = z.string().refine(
```
Comment (juliettech13) [inline src/lib/schemas.ts:23]: lets add some typing here.

Comment (juliettech13) [reply src/lib/schemas.ts:23]: and to all variables below

File: src/lib/schemas.ts [hunk 7]
```diff
@@ -1,26 +1,71 @@
...
+});
+
+export const buildMetadataSchema = z.object({
+  ui: z.record(z.unknown()),
+  change: z.string(),
+  pluginSetup: pluginSetupSchema,
+});
+
+export const subdomainSchema = z.string().regex(/^[a-z0-9-]+$/i, {
+  message: 'Subdomains can only contain alphanumeric characters and dashes.',
```
Comment (juliettech13) [inline src/lib/schemas.ts:70]: "Subdomains can only contain alphanumeric characters and dashes. All letters should be lower cased."

File: src/lib/validators.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import { Address, BuildMetadata, ReleaseMetadata } from 'src/types';
+import { z } from 'zod';
+import { exitWithMessage, strings } from './constants';
+import { getPrivateKey } from './keys';
+import {
```
Comment (juliettech13) [inline src/lib/validators.ts:5]: let's organize these in alphabetic order so they're easier to read through later

File: src/lib/validators.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ */
+export const validateSubdomain = (subdomain: string | undefined): void => {
+  if (typeof subdomain === 'undefined') return;
+
+  const res = subdomainSchema.safeParse(subdomain);
+
+  if (!res.success) exitWithMessage(res.error.message);
+};
+
+export const validateNetwork = (network: AllowedNetworks | undefined) => {
```
Comment (juliettech13) [inline src/lib/validators.ts:75]: Missing docs

File: src/lib/schemas.ts (hunk 6 above)
Comment (pythonpete32) [reply src/lib/schemas.ts:23]: honestly I think its best to leave it to zod here. One of the main benefits of using zod here other that the error handling is for the type infrence. For these the type is easy, but some of the others are massive. 

we dont loose any of the typescript magic by not explisitly using the type here

File: src/lib/schemas.ts [hunk 10]
```diff
@@ -1,26 +1,71 @@
...
+  internalType: z.string(),
+  description: z.string().optional(),
+});
+
+const prepareSchema = z.object({
+  description: z.string().optional(),
+  inputs: z.array(inputSchema).optional(),
+});
+
+const pluginSetupSchema = z.object({
```
Comment (pythonpete32) [inline src/lib/schemas.ts:57]: @juliettech13

but this type will look like this 

```typescript
const pluginSetupSchema: z.ZodObject<{
    prepareInstallation: z.ZodOptional<z.ZodObject<{
        description: z.ZodOptional<ZodString>;
        inputs: z.ZodOptional<z.ZodArray<z.ZodObject<{
            name: ZodString;
            type: ZodString;
            internalType: ZodString;
            description: z.ZodOptional<...>;
        }, "strip", z.ZodTypeAny, {
            ...;
        }, {
            ...;
        }>, "many">>;
    }, "strip", z.ZodTypeAny, {
        ...;
    }, {
        ...;
    }>>;
    prepareUpdate: z.ZodOptional<...>;
    prepareUninstallation: z.ZodOptional<...>;
}, "strip", z.ZodTypeAny, {
    ...;
}, {
    ...;
}>
```

File: src/tests/validators.test.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+  it('exits for invalid address', () => {
+    const mockExitWithMessage = vi.fn();
+    vi.spyOn(console, 'log');
+
+    // mock exitWithMessage
+    vi.spyOn(constants, 'exitWithMessage').mockImplementation(mockExitWithMessage);
+
+    validateAddress('invalid');
+
```
Comment (juliettech13) [inline src/tests/validators.test.ts:35]: remove extra blank line

File: src/tests/validators.test.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+describe('validatePrivateKey', () => {
+  it('exits if private key does not exist', async () => {
+    const mockExitWithMessage = vi.fn();
+    vi.spyOn(constants, 'exitWithMessage').mockImplementation(mockExitWithMessage);
+
+    // mock getPrivateKey to return undefined
+    vi.spyOn(keys, 'getPrivateKey').mockResolvedValue(null);
+
+    await validatePrivateKey();
+
```
Comment (juliettech13) [inline src/tests/validators.test.ts:55]: remove extra line

File: .github/workflows/main-ci.yml [hunk 13]
```diff
@@ -48,6 +48,7 @@ jobs:
           TENDERLY_PROJECT: ${{ secrets.TENDERLY_PROJECT }}
           TENDERLY_USERNAME: ${{ secrets.TENDERLY_USERNAME }}
           PRIVATE_KEY: ${{ secrets.PRIVATE_KEY }}
+          VITE_WEB_3_STORAGE_KEY: ${{ secrets.VITE_WEB_3_STORAGE_KEY }}
```
[the remark this answers is not in the document]
Comment (pythonpete32) [reply .github/workflows/main-ci.yml:51]: using it as a build tool, its using rollup under the hood so we can use that, tsdx or even tsc. I would much rather use a whole other runtime that supports Typescript out of the box like bun tbh

File: src/lib/web3.ts [hunk 14]
```diff
@@ -120,6 +121,93 @@ export const simulateDeployment = async (
...
+    const iFace = new ethers.utils.Interface(PluginRepoFactory__factory.abi);
+
+    await wallet.sendTransaction({
+      from: wallet.address,
+      to: activeContractsList[network.name].PluginRepoFactory,
+      data: iFace.encodeFunctionData('createPluginRepoWithFirstVersion', [
+        subdomain,
+        setupContract,
+        maintainer,
+        '0x00',
```
[the remark this answers is not in the document]
Comment (pythonpete32) [reply src/lib/web3.ts:146]: The metadata is only published in the case where the user is really deploying the plugin. we can refactor so the metadata is published before asking to simulate but the metadata has no effect on the deployment

[the remark this answers is not in the document]
Comment (juliettech13) [reply src/lib/web3.ts:148]: Fair!

**Review by juliettech13 — APPROVED** [review]

**Pull request merged**